metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchSearchSubmitter.java

```diff
@@ -0,0 +1,138 @@
+
+/**
+ * Responsible for submitting searches to Elasticsearch.
+ */
+public class ElasticsearchSearchSubmitter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  /**
+   * The Elasticsearch client.
+   */
+  private TransportClient client;
+
+  public ElasticsearchSearchSubmitter(TransportClient client) {
+    this.client = client;
+  }
+
+  /**
+   * Submit a search to Elasticsearch.
+   * @param request A search request.
+   * @return The search response.
+   */
+  public SearchResponse submitSearch(SearchRequest request) throws InvalidSearchException {
+    LOG.debug("About to submit a search; request={}", ElasticsearchUtils.toJSON(request));
+
+    // submit the search request
+    org.elasticsearch.action.search.SearchResponse esResponse;
+    try {
+      esResponse = client
+              .search(request)
+              .actionGet();
+      LOG.debug("Got Elasticsearch response; response={}", esResponse.toString());
+
+    } catch (SearchPhaseExecutionException e) {
+      String msg = String.format(
+              "Failed to execute search; error='%s', search='%s'",
+              ExceptionUtils.getRootCauseMessage(e),
+              ElasticsearchUtils.toJSON(request));
+      LOG.error(msg, e);
+      throw new InvalidSearchException(msg, e);
+    }
+
+    // check for shard failures
+    if(esResponse.getFailedShards() > 0) {
+      handleShardFailures(request, esResponse);
+    }
+
+    // validate the response status
+    if(RestStatus.OK == esResponse.status()) {
+      return esResponse;
+
+    } else {
+      // the search was not successful
+      String msg = String.format(
+              "Bad search response; status=%s, timeout=%s, terminatedEarly=%s",
+              esResponse.status(), esResponse.isTimedOut(), esResponse.isTerminatedEarly());
+      LOG.error(msg);
+      throw new InvalidSearchException(msg);
+    }
+  }
+
+  /**
+   * Handle individual shard failures that can occur even when the response is OK.  These
+   * can indicate misconfiguration of the search indices.
+   * @param request The search request.
+   * @param response  The search response.
+   */
+  private void handleShardFailures(
+          org.elasticsearch.action.search.SearchRequest request,
+          org.elasticsearch.action.search.SearchResponse response) {
+    /*
+     * shard failures are only logged.  the search itself is not failed.  this approach
+     * assumes that a user is interested in partial search results, even if the
+     * entire search result set cannot be produced.
+     *
+     * for example, assume the user adds an additional sensor and the telemetry
+     * is indexed into a new search index.  if that search index is misconfigured,
+     * it can result in partial shard failures.  rather than failing the entire search,
+     * we log the error and allow the results to be returned from shards that
+     * are correctly configured.
```

Comment:
Hopefully this explanation makes sense.  When there are partial shard failures, we do not fail fast.  We log the problem and return partial results to the user.  I am torn as to whether this is the right approach.